patches/server/1053-Lazily-create-LootContext-for-criterions.patch

```diff
@@ -0,0 +1,33 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: MrPowerGamerBR <git@mrpowergamerbr.com>
+Date: Tue, 21 Nov 2023 12:16:39 -0300
+Subject: [PATCH] Lazily create LootContext for criterions
+
+For each player on each tick, enter block triggers are invoked, and these create loot contexts that are promptly thrown away since the trigger doesn't pass the predicate
+
+To avoid this, we now lazily create the LootContext if the criterion passes the predicate AND if any of the listener triggers require a loot context instance
+
+diff --git a/src/main/java/net/minecraft/advancements/critereon/SimpleCriterionTrigger.java b/src/main/java/net/minecraft/advancements/critereon/SimpleCriterionTrigger.java
+index f0367a9cce13ef576fbb7023c0aba6eb48963606..bbdbf62a7696640d92a6bb758052abaaa6f86edd 100644
+--- a/src/main/java/net/minecraft/advancements/critereon/SimpleCriterionTrigger.java
++++ b/src/main/java/net/minecraft/advancements/critereon/SimpleCriterionTrigger.java
+@@ -54,13 +54,18 @@ public abstract class SimpleCriterionTrigger<T extends SimpleCriterionTrigger.Si
+         PlayerAdvancements playerAdvancements = player.getAdvancements();
+         Set<CriterionTrigger.Listener<T>> set = (Set) playerAdvancements.criterionData.get(this); // Paper - fix AdvancementDataPlayer leak
+         if (set != null && !set.isEmpty()) {
+-            LootContext lootContext = EntityPredicate.createContext(player, player);
++            LootContext lootContext = null; // EntityPredicate.createContext(player, player); // Paper - lazily create LootContext for criterions
+             List<CriterionTrigger.Listener<T>> list = null;
+ 
+             for(CriterionTrigger.Listener<T> listener : set) {
+                 T simpleInstance = listener.trigger();
+                 if (predicate.test(simpleInstance)) {
+                     Optional<ContextAwarePredicate> optional = simpleInstance.playerPredicate();
++                    // Paper start - lazily create LootContext for criterions
++                    if (lootContext == null && optional.isPresent()) {
++                        lootContext = EntityPredicate.createContext(player, player);
++                    }
++                    // Paper end
+                     if (optional.isEmpty() || optional.get().matches(lootContext)) {
```

Comment:
I'm not a super huge fan of the way this logic is structured. To be clear, it seems to work, but the `optional.isPresent()` check on the earlier if statement to counter the `optional.isEmpty()` check below might not be the most clear way to do this.

As much as I don't like using the return value of variable assignment, perhaps this is more clear?
```diff
- if (optional.isEmpty() || optional.get().matches(lootContext)) {
+ if (optional.isEmpty() || optional.get().matches(lootContext = (lootContext == null ? EntityPredicate.createContext(player, player) : lootContext))) {
```

Just a suggestion, if we like what is currently here, prob just add a comment on the isPresent check about it.

Reply:
Here's the patch with your suggestions, if other people also agree that this is better, I will commit it directly to the branch :3 

[1053-Lazily-create-LootContext-for-criterions.patch](https://github.com/PaperMC/Paper/files/13464884/1053-Lazily-create-LootContext-for-criterions.patch)